src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,216 @@
+/*-------------------------------------------------------------------------
+ *
+ * dependencies.c
+ *    Commands to create dependencies of an object on all workers.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/dependency.h"
+#include "catalog/objectaddress.h"
+
+#include "distributed/commands.h"
+#include "distributed/connection_management.h"
+#include "distributed/metadata/dependency.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/remote_commands.h"
+#include "distributed/worker_manager.h"
+
+static List * GetDependencyCreateDDLCommands(const ObjectAddress *dependency);
+
+
+/*
+ * EnsureDependenciesExists finds all the dependencies that we support and makes sure
+ * these are available on all workers. If not available they will be created on the
+ * workers via a separate session that will be committed directly so that the objects are
+ * visible to potentially multiple sessions creating the shards.
+ *
+ * Note; only the actual objects are created via a separate session, the local records to
+ * pg_dist_object are created in this session. As a side effect the objects could be
+ * created on the workers without a catalog entry on the coordinator. Updates to the
+ * objects on the coordinator are not propagated to the workers until the record is
+ * visible on the coordinator.
+ *
+ * This is solved by creating the dependencies in an idempotent manner, either via
+ * postgres native CREATE IF NOT EXISTS, or citus helper functions.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+
+	/* local variables to work with dependencies */
+	List *dependencies = NIL;
+	ListCell *dependencyCell = NULL;
+
+	/* local variables to collect ddl commands */
+	List *ddlCommands = NULL;
+
+	/* local variables to work with worker nodes */
+	List *workerNodeList = NULL;
+	ListCell *workerNodeCell = NULL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	dependencies = GetDependenciesForObject(target);
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		ddlCommands = list_concat(ddlCommands,
+								  GetDependencyCreateDDLCommands(dependency));
+	}
+	if (list_length(ddlCommands) <= 0)
+	{
+		/* no ddl commands to be executed */
+		return;
+	}
+
+	/*
+	 * collect and connect to all applicable nodes
+	 */
+	workerNodeList = ActivePrimaryNodeList();
+	if (list_length(workerNodeList) <= 0)
+	{
+		/* no nodes to execute on */
+		return;
```

Comment:
Returning here means we do not distribute types that were created prior to master_add_node. That might be ok, but I think you can create append/range-distributed tables before you have worker nodes, which means the type propagation will never happen for those tables.

Reply:
I will need to test the case you have described.
It might be I have overlooked the case where there were no workers. If we don't return here it will all be a no-op to the bottom where we mark the objects distributed, that might solve the case you are describing here

Reply:
Yep, not returning here could work, though I'd prefer moving the MarkObjectDistributed calls up here per the other comment.

Reply:
`MarkObjectDistributed` is now moved up. It also is implemented with a true upsert guarding with concurrency issues on `CREATE ... IF NOT EXISTS`.